parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,546 @@
+  final byte[] metadata;
+  /**
+   * The starting index into `value` where the variant value starts. This is used to avoid copying
+   * the value binary when reading a sub-variant in the array/object element.
+   */
+  final int pos;
+
+  /**
+   * The threshold to switch from linear search to binary search when looking up a field by key in
+   * an object. This is a performance optimization to avoid the overhead of binary search for a
+   * short list.
+   */
+  static final int BINARY_SEARCH_THRESHOLD = 32;
+
+  static final ZoneId UTC = ZoneId.of("UTC");
+
+  public Variant(byte[] value, byte[] metadata) {
+    this(value, metadata, 0);
+  }
+
+  Variant(byte[] value, byte[] metadata, int pos) {
+    this.value = value;
+    this.metadata = metadata;
+    this.pos = pos;
+    // There is currently only one allowed version.
+    if (metadata.length < 1 || (metadata[0] & VariantUtil.VERSION_MASK) != VariantUtil.VERSION) {
+      throw VariantUtil.malformedVariant(String.format(
+          "Unsupported variant metadata version: %02X", metadata[0] & VariantUtil.VERSION_MASK));
+    }
+  }
+
+  public byte[] getValue() {
+    if (pos == 0) {
+      // Position 0 means the entire value is used. Return the original value.
+      return value;
+    }
+    int size = VariantUtil.valueSize(value, pos);
+    VariantUtil.checkIndex(pos + size - 1, value.length);
+    return Arrays.copyOfRange(value, pos, pos + size);
+  }
+
+  public byte[] getMetadata() {
+    return metadata;
+  }
+
+  /**
+   * @return the boolean value
+   */
+  public boolean getBoolean() {
+    return VariantUtil.getBoolean(value, pos);
+  }
+
+  /**
+   * @return the long value
+   */
+  public long getLong() {
+    return VariantUtil.getLong(value, pos);
+  }
+
+  /**
+   * @return the double value
+   */
+  public double getDouble() {
+    return VariantUtil.getDouble(value, pos);
+  }
+
+  /**
+   * @return the decimal value
+   */
+  public BigDecimal getDecimal() {
+    return VariantUtil.getDecimal(value, pos);
+  }
+
+  /**
+   * @return the float value
+   */
+  public float getFloat() {
+    return VariantUtil.getFloat(value, pos);
+  }
+
+  /**
+   * @return the binary value
+   */
+  public byte[] getBinary() {
+    return VariantUtil.getBinary(value, pos);
+  }
+
+  /**
+   * @return the UUID value
+   */
+  public UUID getUUID() {
+    return VariantUtil.getUUID(value, pos);
+  }
+
+  /**
+   * @return the string value
+   */
+  public String getString() {
+    return VariantUtil.getString(value, pos);
+  }
+
+  /**
+   * @return the primitive type id from a variant value
+   */
+  public int getPrimitiveTypeId() {
+    return VariantUtil.getPrimitiveTypeId(value, pos);
+  }
+
+  /**
+   * @return the type of the variant value
+   */
+  public VariantUtil.Type getType() {
+    return VariantUtil.getType(value, pos);
+  }
+
+  /**
+   * @return the number of object fields in the variant. `getType()` must be `Type.OBJECT`.
+   */
+  public int objectSize() {
```

Comment:
The spec uses `num_elements` for this. I think it would be better to be clear because `objectSize` could return a length in bytes rather than number of fields or elements.